core/src/ast/macros.rs

```diff
@@ -0,0 +1,279 @@
+        braced!(arm_body in input);
+
+        let body = arm_body.parse::<TokenStream>()?;
+
+        let _semicolon = input.parse::<Token![;]>()?;
+
+        if !input.is_empty() {
+            return Err(syn::Error::new(
+                input.span(),
+                "Diplomat does not support macros of more than one arm.",
+            ));
+        }
+
+        // We don't support any other rules, so we ignore them.
+
+        Ok(Self { match_tokens, body })
+    }
+}
+
+impl MacroDef {
+    pub fn validate(input: ItemMacro) -> TokenStream {
+        let r = input.mac.parse_body::<Self>();
+
+        if let Err(e) = r {
+            e.to_compile_error()
+        } else {
+            TokenStream::default()
+        }
+    }
+
+    fn parse_group(&self, matched: &MacroUse, inner: Cursor) -> TokenStream {
+        let mut stream = TokenStream::new();
+
+        let mut c = inner;
+        while let Some((tt, next)) = c.token_tree() {
+            match &tt {
+                TokenTree::Punct(p) if p.as_char() == '$' => {
+                    if let Some((tt, next)) = next.token_tree() {
+                        if let TokenTree::Ident(i) = tt {
+                            let arg = self.match_tokens.iter().position(|mi| mi.ident == i);
+                            matched.args
+                                [arg.unwrap_or_else(|| panic!("Could not find arg ${i:?}"))]
+                            .to_tokens(&mut stream);
+                            c = next;
+                        } else {
+                            panic!("Expected ident next to $, got {tt:?}");
+                        }
+                    } else {
+                        panic!("Expected token tree.");
+                    }
+                }
+                TokenTree::Group(g) => {
+                    let (inner, _, next) = c.group(g.delimiter()).unwrap();
+                    let group =
+                        proc_macro2::Group::new(g.delimiter(), self.parse_group(matched, inner));
+                    // Once we detect a group, we push it to the array for syn to evaluate.
+                    stream.append(group);
+                    c = next;
+                }
+                _ => {
+                    stream.append(tt);
+                    c = next
+                }
+            }
+        }
+
+        stream
+    }
+
+    fn evaluate_buf(&self, matched: MacroUse) -> TokenStream {
+        let mut stream = TokenStream::new();
+
+        let buf = TokenBuffer::new2(self.body.clone());
+        let mut c = buf.begin();
+        // Search until we find a token to replace:
+        while let Some((tt, next)) = c.token_tree() {
+            match &tt {
+                TokenTree::Punct(punct) if punct.as_char() == '$' => {
+                    if let Some((tt, next)) = next.token_tree() {
+                        if let TokenTree::Ident(i) = tt {
+                            let arg = self.match_tokens.iter().position(|mi| mi.ident == i);
```

Comment:
issue: We should be storing these match tokens as a map from ident to index, i nthe parsed macro.